Added Perl workaround for CUDA <= 8

w8jcik opened the pull request:
@Rombur @ax3l

Resolves https://github.com/spack/spack/issues/23402

How it was tested:

```sh
$ spack install cuda@8.0.61
$ spack install cuda@8.0.44
$ spack install cuda@7.5.18
$ spack install cuda@6.5.14
```

Previously the installation folder was almost empty, now CUDA files are present.

More details in the linked issue.

w8jcik committed:
Added Perl workaround for CUDA <= 8

w8jcik committed:
Re-wrapped comment

w8jcik commented:
The pipeline is failing but it looks unrelated to my change. At least I don't see any connection looking at the log.

What is the procedure in this case? Is it possible that `develop` branch is currently broken? Maybe the pipeline needs to be rerun.

ax3l commented on var/spack/repos/builtin/packages/cuda/package.py:
```diff
@@ -129,6 +129,11 @@ def determine_version(cls, exe):
+        if self.spec.satisfies('@:8.0.61'):
+            # Perl 5.26 removed current directory from module search path,
+            # CUDA 9 has a fix for this, but CUDA 8 and lower don't.
+            env.set('PERL5LIB', self.stage.source_path)
```
Do you think `env.set` is save or should we `env.append_path`?
```suggestion
            env.append_path('PERL5LIB', self.stage.source_path)
```

w8jcik replied:
Sure, it looks better. Sorry I am not very familiar with Spack's API.

ax3l reviewed (dismissed):
Thank you, this is great!
Just a small question if `env.append_path` might be saver?

ax3l commented on var/spack/repos/builtin/packages/cuda/package.py:
```diff
@@ -167,6 +172,16 @@ def install(self, spec, prefix):
+            # running actual installer. This solution is one of the commonly found
+            # in the Internet, when people try to install CUDA <= 8 manually.
+            arguments = [runfile, '--tar', 'mxvf', './InstallUtils.pm']
+            install_shell = which('sh')
+            install_shell(*arguments)
```
Maybe a better name, since we use `install_shell` below and continue after this block:
```suggestion
            extract_shell = which('sh')
            extract_shell(*arguments)
```

ax3l commented on var/spack/repos/builtin/packages/cuda/package.py:
```diff
@@ -167,6 +172,16 @@ def install(self, spec, prefix):
+            # running actual installer. This solution is one of the commonly found
+            # in the Internet, when people try to install CUDA <= 8 manually.
+            arguments = [runfile, '--tar', 'mxvf', './InstallUtils.pm']
+            install_shell = which('sh')
```
We could use the shell only once (move line 197 up):
```suggestion
```

w8jcik replied:
Done

ax3l commented on var/spack/repos/builtin/packages/cuda/package.py:
```diff
@@ -167,6 +172,16 @@ def install(self, spec, prefix):
             os.makedirs(os.path.join(prefix, "src"))
             os.symlink(includedir, os.path.join(prefix, "include"))
 
+        if self.spec.satisfies('@:8.0.61'):
```
```suggestion
        install_shell = which('sh')

        if self.spec.satisfies('@:8.0.61'):
```

w8jcik replied:
Done

ax3l commented on var/spack/repos/builtin/packages/cuda/package.py:
```diff
@@ -167,6 +172,16 @@ def install(self, spec, prefix):
+            # Perl 5.26 removed current directory from module search path.
+            # We are addressing this by exporting `PERL5LIB` earlier, but for some
+            # reason, it is not enough. One more file needs to be extracted before
+            # running actual installer. This solution is one of the commonly found
+            # in the Internet, when people try to install CUDA <= 8 manually.
```
If you have a stackoverflow link or so, feel free to cite it here
```suggestion
            # running the actual installer. This solution is one of the commonly
            # found on the Internet, when people try to install CUDA <= 8 manually.
```

w8jcik committed:
Proofreading corrections

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

w8jcik committed:
Added a reference

w8jcik committed:
Do not override Perl include path

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

w8jcik committed:
Retrieve shell once

w8jcik commented:
Hello,

@ax3l I have addressed the suggestions from review and retested the changes. It works as far as I can tell. Can this pull request be merged?

w8jcik commented:
@ax3l This Pull request is hanging for a while. Can we merge it?

- I have applied your suggestions
- I run tests again (as in the description at the top)
- Pipeline is passing (depends on day of the week, it is fine right now)

ax3l reviewed (approved):
Sorry for the delay.
Thank you for the fix and the ping!
